### ISSUE TITLE
[no-call-expression] add angular 16 signal support

### ISSUE THREAD

**Issue by DerHerrGammler** [angular-eslint/angular-eslint#1380]
**Description and reproduction of the issue**

With angular 16 they implemented signals and signals must be callkable in html. So it should be a good aproach to ignore signals for the rule `no-call-expression` and allow every expression in templates if they are typed as signals

```JSON
{
  "rules": {
    "@angular-eslint/template/no-call-expression": [
      "error",
      {
        "allowList": [],
      },
    ],
  }
}
```

```HTML
<!-- in this case tree is a angular 16 signal and should be allowed -->
{{ tree() | json }}
```

**Versions**

| package                                   | version |
| ----------------------------------        | ------- |
| `@angular-eslint/eslint-plugin-template`  | `16.0.1` |
| `@angular-eslint/template-parser`         | `16.0.1` |
| `@typescript-eslint/parser`               | `5.58.0` |
| `ESLint`                                  | `8.40.0` |
| `node`                                    | `18.15.0` |

```sh
# Please run `npx ng version` in your project and paste the full output here:

     _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/


Angular CLI: 16.0.1
Node: 18.15.0
Package Manager: npm 9.5.0
OS: win32 x64

Angular: 16.0.1
... animations, cdk, cli, common, compiler, compiler-cli, core
... forms, language-service, material, material-moment-adapter
... platform-browser, platform-browser-dynamic, pwa, router
... service-worker

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1600.1
@angular-devkit/build-angular   16.0.1
@angular-devkit/core            16.0.1
@angular-devkit/schematics      16.0.1
@schematics/angular             16.0.1
ng-packagr                      16.0.1
rxjs                            7.8.0
typescript                      5.0.4


```

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest supported version of the packages and checked my `ng version` output per the instructions given here.

**Comment by hanna-becker** [comment]
I just ran into this issue when using a signal for the first time. I don't want to turn off this linter rule globally, because we ran into serious performance issues in the past from function calls in templates. Since signals are memoized (or at least I hope they are...?), this shouldn't be an issue here.

**Comment by nartc** [comment]
@hanna-becker `signal()` has a default equality function (`Object.is`) if not passed in.

**Comment by OmPr366** [comment]
To ensure optimal code quality, we will maintain the enforcement of the `no-call-expression linting` rule. At the same time, we want to effectively utilize `signals`. What is the  way to achieve both without compromising on standards.

**Comment by JamesHenry** [comment]
Linting Angular templates happens at a syntactic level: angular-eslint templates rules operate exclusively on an AST of the template source.

`signal()` and `someOtherThing()` are indistinguishable syntactically, it is only when the semantics of Angular types get applied that the distinction of signal or not can be made.

Somewhat related: the Angular Team have added native support in the Angular compiler for a check to see if a signal was called appropriately: https://github.com/angular-eslint/angular-eslint/issues/1425#issuecomment-1770680336

The reality is there is not really anything we can do here as it stands. `@angular-eslint/template/no-call-expression` just becomes a less useful rule as signal adoption increases and I think that's ok, the framework is evolving and changing, and so the appropriate lint rules change too.

I'm going to close this as there is genuinely nothing we can do right now.

PS A much, much larger effort if somebody is up for the challenge would be to investigate the possibility of create a parser service infrastructure for ESLint and the Angular compiler which parses both source code and templates and would therefore allow validating semantic information and types. It would have the same trade offs as type checking enabled lint rules in typescript-eslint (speed vs completeness/value)

**Comment by jano-kucera** [comment]
Sufficient would be to ignore by Regex, f.e. I've seen many people use '$' **prefix** for signals, similar to Rxjs suffix.

**Comment by ShaneWhittet** [comment]
The https://github.com/angular-eslint/angular-eslint landing page clearly says that typescript/eslint has provided and angular-eslint is capable of adding type checking to its rules, "Please read this dedicated guide to fully understand lint performance and how it is impacted by rules requiring type information: [./docs/RULES_REQUIRING_TYPE_INFORMATION.md](https://github.com/angular-eslint/angular-eslint/blob/main/docs/RULES_REQUIRING_TYPE_INFORMATION.md)"

Due to angular now having a best practice of binding the template to input/output signals. I suggest this issue should be reopened to handle Angular latest direction going toward No Zone.js. 

I suggest making the rule dynamically detect whether the user has parserOptions set for type checking. If not, simply behave as now. When type checking is provided, check for signal types and ignore them. If that is not desired, then a new option to ignore signals or specified types.

If you do not wish to add this capability, please do not close it simply for lack of desire. Another dev can put a PR into the system to address it.

**Comment by JamesHenry** [comment]
@ShaneWhittet please re-read my explanation above, there seems to be a fundamental misunderstanding of the moving parts here.

typescript-eslint is for linting TypeScript code. Angular templates are not TypeScript code and typescript-eslint cannot help in any way here. We do not have type information for templates, as explained in my comment.

You seem to additionally imply that this is not being resolved due to laziness, which is a strange way to try and get folks to work on things for you for free in their personal time.

Maybe you could create a PR which adds support for this? That would be more than welcome. I think you'll discover along the way exactly what is described in this thread

**Comment by ShaneWhittet** [comment]
Thank you for the context. Another issue with my idea, the template doesn't have a reference to the .ts file. I had assumed to be the same file name, though.
